infrastructure/comms/src/connection/net_address/net_address_with_stats.rs

```diff
@@ -0,0 +1,196 @@
+use crate::connection::NetAddress;
+use chrono::prelude::*;
+use std::cmp::{Ord, Ordering};
+
+const MAX_LATENCY_SAMPLE_COUNT: u64 = 100;
+
+#[derive(Debug, Eq)]
+pub struct NetAddressWithStats {
+    pub net_address: NetAddress,
+    pub last_seen: Option<DateTime<Utc>>,
+    pub connection_attempts: u32,
+    pub rejected_message_count: u32,
+    pub avg_latency: u32,
+    latency_sample_count: u64,
+}
+
+impl NetAddressWithStats {
+    /// Constructs a new net address with usage stats
+    pub fn new(
+        net_address: NetAddress,
+        last_seen: Option<DateTime<Utc>>,
+        connection_attempts: u32,
+        rejected_message_count: u32,
+        avg_latency: u32,
+        latency_sample_count: u64,
+    ) -> NetAddressWithStats
+    {
+        NetAddressWithStats {
+            net_address,
+            last_seen,
+            connection_attempts,
+            rejected_message_count,
+            avg_latency,
+            latency_sample_count,
+        }
+    }
+
+    /// Constructs a new net address with usage stats from a net address
+    pub fn from_net_address(net_address: NetAddress) -> NetAddressWithStats {
+        NetAddressWithStats {
+            net_address,
+            last_seen: None,
+            connection_attempts: 0,
+            rejected_message_count: 0,
+            avg_latency: 0,
+            latency_sample_count: 0,
+        }
+    }
+
+    /// Updates the average latency by including another measured latency sample
+    pub fn update_latency(&mut self, latency_measurement: u32) {
+        self.last_seen = Some(Utc::now());
+        self.avg_latency = (((self.avg_latency as u64 * self.latency_sample_count) + latency_measurement as u64) /
+            (self.latency_sample_count + 1)) as u32;
+        if self.latency_sample_count < MAX_LATENCY_SAMPLE_COUNT {
```

Comment:
I think the logic is fine here, @neonknight64 just comment in here, but as far as I can pickup this calcs a new avg everytime with a weighted old avg value.

Reply:
Only up until `count == MAX_LATENCY_SAMPLE_COUNT`. After that, it looks like latency is added and then divided by a constant, meaning that the ave increases monotonically even if latency is constant.

Reply:
You calculate a new sum every time so it should not increase over time as you multiply the old one out. When you calculate the new ave you use `MAX_LATENCY_SAMPLE_COUNT +1` so you technically use a higher sample count than the cons `MAX_LATENCY_SAMPLE_COUNT`.

You only calc the new count after you calculated your new avg

Reply:
I think it is fine, except if a more complex filter is required. When a new latency measurement is received and the ```latency_sample_count``` is equal or have surpassed  the ```MAX_LATENCY_SAMPLE_COUNT``` then the current ```avg_latency``` is scaled so that the new ```latency_measurement``` only makes a small weighted change to the ```avg_latency```. The previous  ```avg_latency``` will have a weight of ```MAX_LATENCY_SAMPLE_COUNT``` and the new ```latency_measurement``` will have a weight of 1. This will produce similar results to a sliding window average without requiring the storage of the actual samples. Yes SW , it currently calculates the average of ```MAX_LATENCY_SAMPLE_COUNT+1``` as it more elegantly handles the case when ```latency_sample_count=0```

Reply:
Ok as we discussed in DM, it's not a strict moving average, but a weighted average of the last point and the historical average. This is fine, but a short note as to what's happening would be useful.